## 圧縮のアルゴリズム

演習問題

二つの和を計算する'add'という関数を作る

- addという関数

    - aという空のリスト
    - bという空のリスト

    - a + b を行う式
    
    - 結果を表示

In [1]:
def add(a, b):
    
    add = a + b
    return add

In [2]:
add(-3, 4)

1

問題２
四則演算を返す'calc'という関数を作る

- calc という関数
    - +する式
    - -する式
    - ＊する式
    - /する式
    - 数値をいれるリスト
    
    出力
    
    calc(4, 2) -> (6, 2, 8, 2)

In [3]:
def add(num1, num2):
    add = num1+ num2
    
    return add



In [4]:
add(2, 3)

5

In [5]:
def pull(num1, num2):
    pull = num1 - num2
    
    return pull


In [6]:
pull(4, 3)

1

In [7]:
def calc(num1, num2):
    add = num1 + num2
    pull = num1 - num2
    dub = num1 * num2
    sep = num1 / num2

    return (add, pull, dub, sep)

In [8]:
calc(4, 2)

(6, 2, 8, 2.0)

問題３
右と左を入れ替える'exchange'という関数を作る

exchange(2, 3) -> (3, 2)

In [9]:
# def exchange(a, b):

#     return (b, a)

## temporaryを使う

def exchange(a, b):
    tmp = a
    a = b
    b = tmp
    
    return (a, b)

In [10]:
exchange(3,7)

(7, 3)

In [11]:
a = [1, 2, 3]
b = [4, 5, 6]

In [12]:
def add(a, b):
   
    
    ## リスト内足し算
    
    #     X1 = a1 + b1
    #     X2 = a2 + b2
    #     X3 = a3 + b3
    #     add = [X1, X2, X3]
    
    return   [a[0] + b[0], a[1] + b[1],  a[2] + b[2]]

In [13]:
add(a, b)

[5, 7, 9]

In [14]:
## リファクタリング　汎用性を高める

- 空のリスト定義　results
    - i = 0,1,2
        - リストa[i]とリストb[i]を足し算
        - resultに結果を追加

In [15]:
a = [1, 2, 3]
b = [4, 5, 6]

In [16]:
def add (a, b):
    results = []
    
    ## リストをa回計算(len)を使う
    for i in range(len(a)): 
        
        ## リストを足し算
        results.append(a[i] + b[i])
        
    return results

In [17]:
add(a, c)

NameError: name 'c' is not defined

 ## 可読性を考える


In [18]:
a = [1, 2, 3]
b = [4, 5, 6]

In [19]:
 results = []
    
#     for i in range(len(a)): 
#         results.append(a[i] + b[i])

## このように書くと、リストaから1つ取り出してxに入れる、リストbから1つ取り出してyに入れる処理になる（zip関数）

for (_a, _b) in zip(a, b):
    results.append(_a + _b)
    

In [20]:
results

[5, 7, 9]

関数化

In [21]:
def add(a, b):
    results = []
    for (_a, _b) in zip(a, b):
        results.append(_a + _b)
        
    return results

In [22]:
add(a, b)

[5, 7, 9]

問題５

- a = [1, 2, 3] 
- b = [4, 5, 6] 
- c = [7, 8, 9] 
- d = [1, 3, 5, 7] 

    - add(a, b) -> [5, 7, 9]
    - add(c, d) -> error
    例外処理を作る

## まず動かしてみる

In [23]:
c = [1, 2, 3,]
d = [1, 2, 3, 4]


In [24]:
def add(c, d):
    if len(c) == len(d):
        results = []
        for (_c, _d) in zip(c, d):
            results.append(_c + _d)
        return results
    else:
            return 'errorだよー'

In [25]:
add(c, d)

'errorだよー'

## 演習課題：　圧縮のアルゴリズム

In [104]:
with open('original.txt', 'r') as f:
    text = f.read()

In [105]:
text[:1000]

'000000000000000000000000000000000000001111111111111000000000000000000000000000000000000000000000000000000000000000000000000011111111110000000000000000000000000000000000000000000000000000000000000000000000000000111111000000000000000000000000000000000000000000000000000000000000000000000000000000001111111111111111111111111111111111111111111111111111111111111111100000000000000000110000000000000000000000000000000000000000000000000000000000000000000000000000011111111111111111111111111111111111111111111111111111111111111111111111100000001111111111111111111111111100000000000000000000000000000000000000000000000000011111111111111111111100000000000000000001111111111111111111111111111111111111111111111111111111111111111111111111111111111111000000000000111111111111111111111111111110000000000000000000000000000001111111111111110000000000000000000000000000000000000000000000000001111111111111111111111111111111111111111111111111111111111111111111110000000000000000000000000000000000000000000000000000000

条件

- 0と１が交互に並んでいる
- 連続して並ぶ回数は１〜９９回まで

In [106]:
len(text)

999536

問題

このtextを情報が「劣化」しないように圧縮する

＊劣化しない＝元に戻した時に、同じ状態になること

- 圧縮を行う関数　encode を作成。
    - text_encode <= 極力文字の少ない
- 複合を行う関数　decode を作成。
    - text_decode == text
    

手元で小さく考える

 - text = 000000111111111111
     - 000000 -> 0[06]
     - 111111111111 -> 1[12]
     
 のように考える。

これを元にエンコード

        - text_encode = 006112...のような形になる。

0 or 1 の２択のため、　006(1)12の（１）は省略できてしまう。

        - text_encode = 00612 になる

- text_encode = encode(text)
- text_encode → '00612'

## decode

- text_decode = decode(text_encode)
- text_decode →　'000000111111111111'

の形になればいい。





## encodeのアルゴリズム

- 先頭が０か1かを取得。
- 何文字続くか、カウントを取得
- １桁の場合は０を入れて補正。（０６のような感じにする。）
- 整数を文字列に変える。
- 全てを文字として足し合わせる。

## デモ

In [107]:
text = '000000111111111111'

In [108]:
# # 先頭の文字を取得
# head = text[0]
# # 先頭の文字を定義
# head

In [109]:
# # for文で文字文を回せる・・・ch(character)
# counts = []
# count = 0
# for ch in text:
    
#     # もしキャラクターが***の場合,カウントを１足します。
#     if ch == head:
#         count += 1
#     else:
#         counts.append(count)## カウントした０がいくつあるかを表示し、終了後カウント初期化   
#         count = 1 ## 1文字目が入っているので、１からスタート
        
#         ## '0'でなければ'１'が入っている。
#         head = ch
        
# ## for文に対してelseを使う（python独特のもの。意外と便利）　繰り返しの最後に処理
# else:
#     counts.append(count)

In [110]:
counts

[6, 12]

関数化

In [111]:
def get_count(text):
    
    _head = text[0]
    head = _head
    
    counts = []
    count = 0
    for ch in text:
        if ch == head:
            count += 1
        else:
            counts.append(count) 
            count = 1
            head = ch
    else:
        counts.append(count)
        
    return _head, counts

In [112]:
head, counts = get_count(text)

In [113]:
head

'0'

In [114]:
counts

[6, 12]

## 整数から文字列に変える関数を作る。
- 整数'6'を'06'に変換
- 整数'12'を'12'に変換

In [115]:
count = 6

In [116]:
# if count < 10:
#     print('0' + str(count))
# else:
#     print(str(count))

関数化

In [117]:
def int2str(count):
    if count < 10:
        return ('0' + str(count))
    else:
        return (str(count))

In [118]:
int2str(6)

'06'

## 全てつなげる

In [119]:
count

6

In [120]:
text_encode = head
for count in counts:
    ## 足して更新する
    text_encode += int2str(count)

In [121]:
text_encode

'00612'

In [122]:
def encode(text):
    ## head, countを抜き出す
    head, counts = get_count(text)
    text_encode = head
    for count in counts:
        ## 足して更新する
        text_encode += int2str(count)
    return text_encode

In [123]:
text_encode = encode(text)
text_encode

'00612'

## 本番のデータに適用

In [124]:
with open('original.txt', 'r') as f:
    text = f.read()

In [125]:
len(text)

999536

In [126]:
text_encord = encode(text)

In [127]:
len(text_encord)

40001

In [128]:
text_encode[:100]

'00612'

In [129]:
len(text_encode) / len(text) * 100

0.0005002321076979718

## decode

- text_decode = decode(text_encode)
- text_decode →　'000000111111111111'

    - len(text_encode) -> 40001
    - len(text_decode) -> 999536
    
- 先頭（head）と本文（body）を分ける
- (body)を2文字ずつに分割　→　'0612' →　'06','12'
- 文字列を整数に変換　→　'06'　→　'6'
- headの文字から交互に、カウント分だけ文字を足し合わせる。


In [93]:
# ## 先頭を取得
# head = text_encode[0]
# head

In [94]:
# ## bodyと分ける
# body = text_encode[1:]
# body

In [95]:
# # bodyを分割

# # bodyを2つずつ
# for i in range(0, len(body), 2):
    
#     # iは０番目のため、０〜2番目までを表示
#     num = body[i:i+2]
#     print(num)

In [96]:
# # 文字列を整数に変換　→　intを使う

# for i in range(0, len(body), 2):
    
#     # iは０番目のため、０〜2番目までを表示　int を使い文字列を整数に変換
#     num = int(body[i:i+2])
#     print(num)

In [97]:
#  headの文字から交互にカウント文だけ足し合わせる
# これを利用する

# '0' * 6

In [98]:
# '1' * 12

In [99]:
# head = text_encode[0]
# body = text_encode[1:]
# text_decode = ''

# for i in range(0, len(body), 2):
#     num = int(body[i:i+2])
#     text_decode += head * num
    
#     # 文字列を交互にカウント
#     if head == '0':
#         head = '1'
#     else:
#         head = 0
#         print(text_decode)
    

関数化

In [133]:
def decode(text_encode):   
    head = text_encode[0]
    body = text_encode[1:]
    text_decode = ''

    for i in range(0, len(body), 2):
        num = int(body[i:i+2])
        text_decode += head * num

        # 文字列を交互にカウント
        if head == '0':
            head = '1'
        else:
            head = '0'
            
    return text_decode

In [134]:
text_decode = decode(text_encode)

In [135]:
text_decode

'000000111111111111'

本番データに適用

In [136]:
with open('original.txt', 'r') as f:
    text = f.read()

In [139]:
len(text)

999536

In [140]:
text_encode = encode(text)

In [142]:
len(text_encode)

40001

In [143]:
text_decode = decode(text_encode)

In [145]:
len(text_decode)

999536